jerry-core/ecma/operations/ecma-map-object.c

```diff
@@ -0,0 +1,757 @@
+
+    if (ecma_get_object_type (&map_object_p->header.object) == ECMA_OBJECT_TYPE_CLASS
+        && map_object_p->header.u.class_prop.class_id == (is_weak_map ? LIT_MAGIC_STRING_WEAK_MAP_UL
+                                                                      : LIT_MAGIC_STRING_MAP_UL))
+    {
+      return map_object_p;
+    }
+  }
+
+  ecma_raise_type_error (ECMA_ERR_MSG (is_weak_map ? "Expected a WeakMap object."
+                                                   : "Expected a Map object."));
+  return NULL;
+} /* ecma_op_map_get_object */
+
+/**
+ * Returns with the size of the map object.
+ *
+ * @return size of the map object as ecma-value.
+ */
+ecma_value_t
+ecma_op_map_size (ecma_value_t this_arg) /**< this argument */
+{
+  ecma_map_object_t *map_object_p = ecma_op_map_get_object (this_arg, false);
+  if (map_object_p == NULL)
+  {
+    return ECMA_VALUE_ERROR;
+  }
+
+  return ecma_make_uint32_value (map_object_p->header.u.class_prop.u.length);
+} /* ecma_op_map_size */
+
+/**
+ * Creates an error message for invalid keys.
+ *
+ * @return ECMA_VALUE_ERROR
+ */
+static ecma_value_t
+ecma_op_weak_map_key_error (void)
+{
+  return ecma_raise_type_error (ECMA_ERR_MSG ("WeakMap keys must be objects."));
+}
+
+/**
+ * Linear search for the value in the map storage
+ *
+ * @return pointer to value if key is found
+ *         NULL otherwise
+ */
+static ecma_value_t *
+ecma_builtin_map_search (jmem_cpointer_t first_chunk_cp, /**< first chunk */
+                         ecma_value_t key) /**< key to search */
+{
+  if (JERRY_UNLIKELY (first_chunk_cp == ECMA_NULL_POINTER))
+  {
+    return NULL;
+  }
+
+  ecma_map_object_chunk_t *chunk_p = ECMA_GET_NON_NULL_POINTER (ecma_map_object_chunk_t,
+                                                                first_chunk_cp);
+
+  bool is_direct = true;
+  const ecma_string_t *key_str_p = NULL;
+  ecma_number_t key_float = 0;
+
+  if (ecma_is_value_non_direct_string (key))
+  {
+    key_str_p = ecma_get_string_from_value (key);
+    is_direct = false;
+  }
+  else if (ecma_is_value_float_number (key))
+  {
+    key_float = ecma_get_float_from_value (key);
+    is_direct = false;
+  }
+
+  ecma_value_t *item_p = chunk_p->items;
+  ecma_value_t last_key = ECMA_VALUE_ARRAY_HOLE;
+
+  while (true)
+  {
+    ecma_value_t item = *item_p++;
+
+    if (JERRY_UNLIKELY (item == ECMA_VALUE_ARRAY_HOLE))
+    {
+      JERRY_ASSERT (last_key == ECMA_VALUE_ARRAY_HOLE);
+      continue;
+    }
+
+    if (JERRY_UNLIKELY (ecma_is_value_pointer (item)))
+    {
+      item_p = (ecma_value_t *) ecma_get_pointer_from_value (item);
+
+      if (item_p == NULL)
+      {
+        JERRY_ASSERT (last_key == ECMA_VALUE_ARRAY_HOLE);
+        return NULL;
+      }
+
+      JERRY_ASSERT (!ecma_is_value_pointer (*item_p));
+      continue;
+    }
+
+    if (last_key == ECMA_VALUE_ARRAY_HOLE)
+    {
+      last_key = item;
+    }
+    else
+    {
+      if (JERRY_LIKELY (is_direct))
+      {
+        if (key == last_key)
```

Comment:
These `if` statements can be merged.

Reply:
I don't think so, because if `is_direct` is true, but the `key != last_key`, we fall to the `ecma_is_value_float_number (last_key)` which we shouldn't.